runtime/compiler/env/ClassLoaderTable.cpp

```diff
@@ -182,28 +203,29 @@ TR_PersistentClassLoaderTable::associateClassLoaderWithClass(J9VMThread *vmThrea
    if (info)
       return;
 
-#if defined(J9VM_OPT_JITSERVER)
-   bool useAOTCache = _persistentMemory->getPersistentInfo()->getJITServerUseAOTCache();
-   const J9UTF8 *nameStr = J9ROMCLASS_CLASSNAME(((J9Class *)clazz)->romClass);
-   const uint8_t *name = J9UTF8_DATA(nameStr);
-   uint16_t nameLength = J9UTF8_LENGTH(nameStr);
-#endif /* defined(J9VM_OPT_JITSERVER) */
 
-   void *chain = _sharedCache->rememberClass(clazz);
-   if (!chain)
-      {
-#if defined(J9VM_OPT_JITSERVER)
-      if (useAOTCache && TR::Options::getVerboseOption(TR_VerboseJITServer))
-         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
-            "ERROR: Failed to get class chain for %.*s loaded by %p",
-            nameLength, (const char *)name, loader
-         );
-#endif /* defined(J9VM_OPT_JITSERVER) */
+   void *chain = _sharedCache ? _sharedCache->rememberClass(clazz) : NULL;
+   // If we are using the JITServer AOT cache, we need to remember the name of clazz with or without chain.
+   // If we aren't using it, there is no point in continuing.
+   if (!chain && !useAOTCache)
       return;
+   TR_ASSERT(!_sharedCache || !chain || _sharedCache->isPointerInSharedCache(chain), "Class chain must be in SCC");
+
+   J9UTF8 *nameStr = NULL;
+#if defined(J9VM_OPT_JITSERVER)
+   const uint8_t *name = NULL;
+   size_t nameLength = 0;
+   if (useAOTCache)
+      {
+      auto romClass = ((J9Class *)clazz)->romClass;
+      auto romName = J9ROMCLASS_CLASSNAME(romClass);
+      nameStr = (_sharedCache && _sharedCache->isROMClassInSharedCache(romClass)) ? romName : copyJ9UTF8(romName, _persistentMemory);
+      name = J9UTF8_DATA(romName);
+      nameLength = J9UTF8_LENGTH(romName);
       }
-   TR_ASSERT(_sharedCache->isPointerInSharedCache(chain), "Class chain must be in SCC");
+#endif /* defined(J9VM_OPT_JITSERVER) */
 
-   info = new (_persistentMemory) TR_ClassLoaderInfo(loader, chain);
+   info = new (_persistentMemory) TR_ClassLoaderInfo(loader, chain, nameStr);
```

Comment:
Do we delete such entries from the table? How do we handle the freeing of persistent memory for the name of the class? We much remember whether we used persistent memory, in which case we need to free it, or not, in which case we don't have to touch it.

Reply:
Looking at the code, the "remove" function only detaches an entry from the bucket, but does not delete anything.
I don't remember if there was any technical issue with freeing the memory, but in the future we may want to actually free the memory, including the new name string that this PR has added. Please open a separate issue for it.

Reply:
I should probably check that the name string isn't NULL before freeing it, though.

Reply:
Oh, never mind. That shouldn't be necessary. I forgot that we now track names exactly when `getJITServerUseAOTCache()` is true, and I don't think that will ever switch from false to true during the operation of the class loader table.